### PR TITLE
RPC 0.6.0 cleanup

### LINKED ISSUE #372: Refactor V3 interfaces
### Feature Request

Function signature and naming of account is a bit inconsistent after RPC 0.6.0 update.
Should consider refactoring it to create better API.

### LINKED ISSUE #374: Add missing RPC 0.6.0 tests
### Feature Request

Should fully test features that weren't added in #358

Including:
- Calculating v3 hashes
- Uint64, Uint128, ...
 - v3 network tests
   - estimateFeeV3

### LINKED ISSUE #393: Unify JsonRpcErrors
### Feature Request

Instead of using multiple classes for possible jsonrpc errors, make "data" field a String and accept any raw json.

### LINKED ISSUE #395: `signDeclareV3` version mismatch when used for fee estimate
### Feature Request

`signDeclareV3` with `forFeeEstimate = true` signs transaction using query version, but sends request with a regular version

### LINKED ISSUE #396: `executeV3` estimates incorrect fee
It seems `executeV3` estimates the fee for v1 transaction (in wei), therefore it's generally not enough to execute it later (with fee paid in fri)

### LINKED ISSUE #402: Support deploying using invoke v3 transaction
### Feature Request

In `StandardDeployer`, support deploying contracts using v3 invoke transactions

### PR DESCRIPTION
**Pull request by DelevoXDG**
## Describe your changes

Refactoring and cleanup post #358. Includes missing transaction v3 functionalities, more streamlined interface, bug fixes and tests covering Starknet 0.13 features.

- `Account`, `StandardAccount`
  - all methods that have v1, v2 and v3 variants now have `V1`, `V2`, `V3` postfixes respectively
  - `estimateFeeV(1|3)` no longer accepts `Set<SimulationFlagForEstimateFee>`, instead takes `skipValidate: Bool`
  - Add missing `estimateFeeV3(Call)`
  - #396 
-  `Deployer`, `StandardDeployer`: Add `deployContractV3` methods ; rename `deployContract` ->`deployContractV1` 
- Rename `ExecutionParamsV3` -> `InvokeParamsV3`
- Restrict usage of v3 fields
  - `TransactionPayload`: make primary constructor private; add custom constructor with hardcoded values
  - `TransactionFactory`: don't accept v3 transaction params 
- Add `DeprecatedTransaction` and `TransactionV3` interfaces 
- #393  
- Fix declare v3 payload for fee estimate
- `StarknetChainId`: Rename `SEPOLIA_TESTNET`->`SEPOLIA`, renamed `SEPOLIA_INTEGRATOIN`->`INTEGATION_SEPOLIA` (should we remove enum altogether?)
- Add missing RPC 0.6.0 tests
  - `TransactionHashCalculator`: test v3s; add missing declare v1/v2 tests
  - Rename `FeltTests` -> `NumAsHexBaseTests `; test all numbers; nest Felt-specific tests; add Uint256 tets
  - `StandardAccountTest`
    - Nest test cases: add `DeclareEstimateTest`, `DeclareTests`, `DeployAccountEstimateTest`, `DeployAccountTest`, `DeployAccountEstimateTest`
    - Add missing v3 tests (`simulateTransactions`, `estimateFee`)
    - Add estimate fee with `SKIP_VALIDATE` test
  - `StandardDeployerTest`: test deploy v3
    - Remove `contract_with_constructor.cairo` (no longer used in tests)

## Linked issues

Closes #402
Closes #372 
Closes #393 
Closes #395 
Closes #396 
Closes #374 

## Breaking changes

- [x] This issue contains breaking changes

- `Account`, `StandardAccount`: all methods that have v1, v2 and v3 variants now have `V1`, `V2`, `V3` postfixes respectively
- `Deployer`, `StandardDeployer`: `deployContract` split into `deployContractV1` and `deployContractV3` 
- `estimateFeeV(1|3)` no longer accepts `Set<SimulationFlagForEstimateFee>`, instead takes `skipValidate: Bool`
- Removed transaction v3 parameters (`tip`, `paymasterData`, `accountDeploymentData`, `nonceDataAvailabilityMode`, `feeDataAvailabilityMode`) from `TransactionPayload` constructors
- Removed transaction v3 parameters (`tip`, `paymasterData`, `accountDeploymentData`, `nonceDataAvailabilityMode`, `feeDataAvailabilityMode`) from `TransactionFactory` methods
- Renamed `ExecutionParamsV3` -> `InvokeParamsV3`
- `StarknetChainId`: Renamed `SEPOLIA_TESTNET`->`SEPOLIA`, renamed `SEPOLIA_INTEGRATOIN`->`INTEGATION_SEPOLIA`,

### CODE REVIEW

File: lib/src/main/kotlin/com/swmansion/starknet/data/types/transactions/Transaction.kt [hunk 1]
```diff
@@ -210,11 +210,6 @@ data class InvokeTransactionV3(
             senderAddress = senderAddress,
             version = version,
             resourceBounds = resourceBounds,
-            tip = tip,
-            paymasterData = paymasterData,
-            accountDeploymentData = accountDeploymentData,
-            nonceDataAvailabilityMode = nonceDataAvailabilityMode,
-            feeDataAvailabilityMode = feeDataAvailabilityMode,
```
Comment (DelevoXDG) [inline lib/src/main/kotlin/com/swmansion/starknet/data/types/transactions/Transaction.kt:217]: Should we also make primary constructor private similarly to `TransactionPayload`, v3 params and so on, and create constructor with hardcoded fields? I decided to not do it for now:
1. to not further clutter the pR
2. because there were previously some non-rpc fields and we didn't hardcode anything
3. we mainly use `TransactionFactory` to correctly create `Transaction` objects anyway

This would be easily solved with #394, but we decided to do it separately.

File: lib/src/main/kotlin/com/swmansion/starknet/data/types/transactions/TransactionPayload.kt [hunk 2]
```diff
@@ -199,11 +186,34 @@ data class DeclareTransactionV3Payload(
...
+        signature = signature,
+        senderAddress = senderAddress,
+        compiledClassHash = compiledClassHash,
+        resourceBounds = resourceBounds,
+        tip = Uint64.ZERO,
+        paymasterData = emptyList(),
+        accountDeploymentData = emptyList(),
+        nonceDataAvailabilityMode = DAMode.L1,
+        feeDataAvailabilityMode = DAMode.L1,
+        version = version,
```
Comment (DelevoXDG) [inline lib/src/main/kotlin/com/swmansion/starknet/data/types/transactions/TransactionPayload.kt:214]: Should I change this to:
```suggestion
        tip = TransactionFactory.tip,
        paymasterData = TransactionFactory.paymasterData,
        accountDeploymentData = TransactionFactory.accountDeploymentData,
        nonceDataAvailabilityMode = TransactionFactory.nonceDataAvailabilityMode,
        feeDataAvailabilityMode = TransactionFactory.feeDataAvailabilityMode,
```
I don't like the idea of relying on another object, but I don't like defaults scattered around also

File: lib/src/main/kotlin/com/swmansion/starknet/data/types/transactions/Transaction.kt [hunk 3]
```diff
@@ -146,21 +170,6 @@ data class InvokeTransactionV1(
             version = version,
         )
     }
-
-    companion object {
```
Comment (THenry14) [inline lib/src/main/kotlin/com/swmansion/starknet/data/types/transactions/Transaction.kt:150]: is this no longer needed?

File: lib/src/main/kotlin/com/swmansion/starknet/data/types/transactions/Transaction.kt (hunk 1 above)
Comment (THenry14) [reply lib/src/main/kotlin/com/swmansion/starknet/data/types/transactions/Transaction.kt:217]: no need, I suppose

File: lib/src/main/kotlin/com/swmansion/starknet/data/types/transactions/TransactionPayload.kt (hunk 2 above)
Comment (THenry14) [reply lib/src/main/kotlin/com/swmansion/starknet/data/types/transactions/TransactionPayload.kt:214]: I'd leave this as is right now tbh.

File: lib/src/main/kotlin/com/swmansion/starknet/provider/rpc/BuildJsonHttpDeserializer.kt [hunk 4]
```diff
@@ -21,54 +20,20 @@ private data class JsonRpcResponse<T>(
...
-internal data class JsonRpcContractErrorData(
-    @SerialName("revert_error")
-    val revertError: String,
-) {
-    override fun toString(): String {
-        return revertError
-    }
-}
+    val data: String? = null,
+)
```
Comment (THenry14) [inline lib/src/main/kotlin/com/swmansion/starknet/provider/rpc/BuildJsonHttpDeserializer.kt:36]: what is this change related to?

**Review by THenry14 — COMMENTED** [review]
glanced through tests
let's remember about carefully describing this changes in changelog, when new version is released

File: lib/src/main/kotlin/com/swmansion/starknet/data/types/transactions/Transaction.kt (hunk 3 above)
Comment (DelevoXDG) [reply lib/src/main/kotlin/com/swmansion/starknet/data/types/transactions/Transaction.kt:150]: We don't use it anywhere, I'm not sure if we ever did use it. I couldn't find it being used in #197 (PR where it was added). Since this is an internal method and we don't use it anywhere internally, I think we don't need it.

**Review by THenry14 — APPROVED** [review]
looks good, I guess 😅
